application/common/components/MfaBackendWebAuthn.php

```diff
@@ -0,0 +1,187 @@
+
+        return $this->client->webauthnCreateRegistration($headers);
+    }
+
+    /**
+     * Initialize authentication sequence
+     * @param int $mfaId The MFA ID
+     * @param string $rpOrigin The Replay Party Origin URL (with scheme, without port or path)
+     * @return array
+     * @throws NotFoundHttpException
+     * @throws GuzzleException
+     */
+    public function authInit(int $mfaId, string $rpOrigin = ''): array
+    {
+        $mfa = Mfa::findOne(['id' => $mfaId]);
+        if ($mfa == null) {
+            throw new NotFoundHttpException("MFA record for given ID not found");
+        }
+
+        $headers = $this->getWebAuthnHeaders(
+            $mfa->user->username,
+            $mfa->user->getDisplayName(),
+            $rpOrigin,
+            $mfa->external_uuid
+        );
+
+        return $this->client->webauthnCreateAuthentication($headers);
+    }
+
+    /**
+     * Verify response from user is correct for the MFA backend device
+     * @param int $mfaId The MFA ID
+     * @param string $value The stringified JSON response from the browser credential api
+     * @param string $rpOrigin The Replay Party Origin URL (with scheme, without port or path)
+     * @return bool
+     * @throws GuzzleException
+     */
+    public function verify(int $mfaId, string $value, string $rpOrigin = ''): bool
+    {
+        $mfa = Mfa::findOne(['id' => $mfaId]);
+        if ($mfa == null) {
+            throw new NotFoundHttpException("MFA record for given ID not found");
+        }
+
+        $headers = $this->getWebAuthnHeaders(
+            $mfa->user->username,
+            $mfa->user->getDisplayName(),
+            $rpOrigin,
+            $mfa->external_uuid
+        );
+
+        $data = Json::decode($value);
+        if ($data == null) {
+            throw new ServerErrorHttpException("Missing data or unable to decode as JSON", 1638447364);
+        }
+
+        if ($mfa->verified === 1) {
+            return $this->client->webauthnValidateAuthentication($headers, $data);
+        } else {
+            if ($this->client->webauthnValidateRegistration($headers, $data)) {
+                $mfa->verified = 1;
+                if (! $mfa->save()) {
+                    throw new ServerErrorHttpException(
+                        "Unable to save WebAuthn record after verification. Error: " . print_r($mfa->getFirstErrors(), true)
+                    );
+                }
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Delete WebAuthn configuration
+     * @param int $mfaId
+     * @return bool
+     * @throws NotFoundHttpException
+     * @throws GuzzleException
+     */
+    public function delete(int $mfaId): bool
+    {
+        $mfa = Mfa::findOne(['id' => $mfaId]);
+        if ($mfa == null) {
+            throw new NotFoundHttpException("MFA record for given ID not found");
+        }
+
+        if (is_string($mfa->external_uuid)) {
```

Comment:
If it's not a string, should something different happen?

Reply:
good question, I guess I copied it from the u2f method without considering it. I think we actually just want to check that it's not empty, not that it's not a string. I'll update it